simple-generated/src/event.rs

```diff
@@ -1,87 +1,116 @@
-                            CrosstermEvent::FocusGained => Ok(()),
-                            CrosstermEvent::FocusLost => Ok(()),
-                            CrosstermEvent::Paste(_) => unimplemented!(),
-                        }
-                        .expect("failed to send terminal event")
-                    }
-
-                    if last_tick.elapsed() >= tick_rate {
-                        sender.send(Event::Tick).expect("failed to send tick event");
-                        last_tick = Instant::now();
-                    }
-                }
-            })
-        };
-        Self {
-            sender,
-            receiver,
-            handler,
-        }
+        thread::spawn(move || event_thread(sender, tick_rate));
+        Self { receiver }
     }
 
     /// Receive the next event from the handler thread.
     ///
-    /// This function will always block the current thread if
-    /// there is no data available and it's possible for more data to be sent.
-    pub fn next(&self) -> AppResult<Event> {
-        Ok(self.receiver.recv()?)
+    /// This function will always block the current thread if there is no data available unless the
+    /// event source has been closed.
+    pub fn next(&self) -> color_eyre::Result<Event> {
+        let event = self.receiver.recv()?;
+        Ok(event)
+    }
+}
+
+/// An event thread that listens for terminal events.
+///
+/// This function is responsible for listening to terminal events and sending them to the main
+/// thread. It sends a tick event at a regular interval. The tick rate is specified by the
+/// `tick_rate` parameter. If an error occurs, it sends an error event to the main thread and then
+/// stops running.
+fn event_thread(sender: mpsc::Sender<Event>, tick_rate: Duration) {
+    let mut last_tick: Option<Instant> = None;
+    loop {
+        if last_tick
+            .map(|tick| tick.elapsed() >= tick_rate)
+            .unwrap_or(true)
+        {
+            sender.send(Event::Tick).expect("failed to send tick event");
+            last_tick = Some(Instant::now());
+        }
+
+        let timeout = last_tick.map_or(Duration::ZERO, |tick| {
+            tick_rate.saturating_sub(tick.elapsed())
+        });
+
+        match event::poll(timeout) {
+            Ok(false) => {
+                // no new events waiting
+                continue;
+            }
+            Ok(true) => {}
+            Err(e) => {
+                let _ = sender.send(Event::Error(e));
+                break;
+            }
+        }
+
+        match event::read() {
+            Err(err) => {
+                let _ = sender.send(Event::Error(err));
+                break;
+            }
+            Ok(event) => match event {
+                CrosstermEvent::Key(e) if e.kind == KeyEventKind::Press => {
+                    // ignore key release / repeat events
+                    let _ = sender.send(Event::Key(e));
```

Comment:
Send failures are expected and just indicate that the receiving side is closed. This is not an error that should cause a panic, it should gracefully just stop (perhaps the tick event not being able to be sent should just break out of the loop).